teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -0,0 +1,205 @@
+Final tasks are perfectly suited for any cleanup operations that have to be performed when execution is completed.
+That is required in case of success, failure, as well as pipeline run cancellation.
+ 
+Currently, there is no way to terminate a `PipelineRun` gracefully.
+The existing cancellation capability is problematic in the real use cases.
+
+This is especially important for users of [Kubeflow Pipelines with Tekton backend](https://github.com/kubeflow/kfp-tekton/).
+Kubeflow Pipelines supports [exit handler](https://www.kubeflow.org/docs/components/pipelines/overview/pipelines-overview/)
+which guarantees that selected operations are triggered whenever pipeline run is completed.
+Those actions are executed in case of a pipeline run cancellation as well. 
+Given the fact that Kubeflow Pipelines' `ExitHandler` is implemented in Tekton using final tasks,
+there is a significant inconsistency in the describe behaviour.
+
+Lastly, final tasks should be triggered on a pipeline run timeout, which is a standard error scenario.
+Running final tasks infinitely should be prevented with the additional configuration of a finalization timeout.
+There is a separate proposal: [TEP-0047](https://github.com/tektoncd/community/pull/326) that covers this part.
+
+Related issues:
+- https://github.com/kubeflow/kfp-tekton/issues/506
+- https://github.com/tektoncd/community/pull/326
+
+
+### Goals
+
+- Add support for graceful a pipeline run termination / stop, that would wait for final tasks to be completed before
+  termination.
+
+### Non-Goals
+
+- There is not intention to change the existing pipeline run cancellation behaviour, 
+  but rather provide an alternative one that would support graceful run termination.
+
+### Use Cases (optional)
+
+- As a Kubeflow user, I run a training pipeline that spawns a [`TFJob`](https://www.kubeflow.org/docs/components/training/tftraining/).
+  Once the pipeline execution is stopped the training job should be terminated to limit resource usage.
+- As a CodeEngine user, I run a pipeline that [submits a batch job](https://cloud.ibm.com/docs/codeengine?topic=codeengine-cli#cli-jobrun).
+  Once the pipeline execution is stopped the batch job should be terminated to limit the service costs.
+- As a kubernetes user, I run a pipeline that provisions a new cluster.
+  The pipeline is executed and stopped. The k8s cluster resource should be freed up.
+
+## Requirements
+
+- Users should be able to gracefully stop a pipeline run and cleanup external resources. 
+
+## Proposal
+
+To gracefully terminate (stop) a `PipelineRun` that's currently executing, users update its definition to mark it as stopped. 
+When you do so, the spawned `TaskRuns` are also marked as stopped and all associated Pods are deleted.
+In parallel the final tasks are triggered.
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: go-example-git
+spec:
+  # […]
+  status: "PipelineRunStopped"
```

Comment:
We do have an internal state `PipelineRunStopping` which is triggered when a `taskRun` fails and `pipelineRun` is waiting for the other running `taskRuns` to finish. This state also does not stop `finally` to be scheduled. We could expose the same state and allow users to configure it to stop `tasks` section.

Reply:
Oh, it would cause a conflict, since here we would want to stop the running `taskRuns` and not wait until they finish 🙃 .

Reply:
@pritidesai exposing `PipelineRunStopping` could potentially solve the new requirement to optionally wait for running tasks. But that would probably require a little more discussion.